src/parallel_progress.jl

```diff
@@ -0,0 +1,364 @@
+
+                if f == PP_NEXT
+                    if count_overshoot || !has_finished(progresses[p])
+                        next!(progresses[p], args...; kwt..., keep=false)
+                        !count_finishes && next!(mainprogress; keep=false)
+                    end
+                else
+                    prev_p_counter = progresses[p].counter
+                    
+                    if f == PP_FINISH
+                        finish!(progresses[p], args...; kwt..., keep=false)
+                    elseif f == PP_CANCEL
+                        finish!(progresses[p]; keep=false)
+                        cancel(progresses[p], args...; kwt..., keep=false)
+                    elseif f == PP_UPDATE
+                        if !isempty(args)
+                            value = args[1]
+                            !count_overshoot && progresses[p] isa Progress && (value = min(value, progresses[p].n))
+                            update!(progresses[p], value, args[2:end]...; kwt..., keep=false)
+                        else
+                            update!(progresses[p]; kwt..., keep=false)
+                        end
+                    end
+
+                    !count_finishes && update!(mainprogress, 
+                        mainprogress.counter - prev_p_counter + progresses[p].counter; keep=false)
+                end
+
+                if !already_finished && has_finished(progresses[p])
+                    delete!(taken_offsets, progresses[p].offset)
+                    count_finishes && next!(mainprogress; keep=false)
+                end
+            end
+        end
+    catch err
+        # channel closed should only happen from Base.close(mp), which isn't an error
+        if err != Base.closed_exception()
+            bt = catch_backtrace()
+            println()
+            showerror(stderr, err, bt)
+            println()
+        end
+    finally
+        print("\n" ^ (max_offsets+1))
+        close(mp)
+    end
+end
+
+"""
+    close(p::Union{ParallelProgress,MultipleProgress})
+
+empties and closes the channel of the progress and replaces it with a `FakeChannel` to allow overshoot
+
+"""
+function Base.close(p::Union{ParallelProgress,MultipleProgress})
+    channel = p.channel
+    p.channel = FakeChannel()
+    while isready(channel) # empty channel to avoid waiting `put!`
+        take!(channel)
+    end
+    close(channel)
+end
+
+has_finished(p::Progress) = p.counter >= p.n
+has_finished(p::ProgressThresh) = p.triggered
+has_finished(p::ProgressUnknown) = p.done
+has_finished(p::ParallelProgress) = isfakechannel(p.channel)
+has_finished(p::MultipleProgress) = isfakechannel(p.channel)
+
+isfakechannel(_) = false
+isfakechannel(::FakeChannel) = true
+isfakechannel(mc::MultipleChannel) = isfakechannel(mc.channel)
+
+"""
+    addprogress!(mp[i], T::Type{<:AbstractProgress}, args...; kw...)
+
+will add the progressbar `T(args..., kw...)` to the MultipleProgress `mp` at index `i`
```

Comment:
To be clear, if there was already a progress bar `mp[i]`, what happens to it?  Does it get shifted to `i+1`?

Reply:
It's a bit confusing that you modify `mp` by passing one of its *elements* to this function. 

Wouldn't it be clearer to use `insert!` etcetera?  e.g.
```jl
insert!(mp, i, T, args...) # inserts at index i
push!(mp, T, args...) # adds to end
pushfirst!(mp, T, args...) # inserts at beginning
```
?   These semantics would require the return value to be `mp`.   Then you would use `mp[i]`, `mp[end]`, and `mp[begin]` to get the corresponding progress bar.

Reply:
right now the new progressbar would take the place of the previous and if it isn't finished, it would stay stuck. I was thinking of a usecase where each worker has only one task and they can all do `mp[myid()]`

`push!` is probably the best, since we don't want the other progressbars changing index
and/or use `setindex!` or `replace!` to be more explicit that it replaces the previous progressbar

Reply:
concerning `push!`, I was worried two workers would take the same index at the same time, I'm not sure how to make sure it doesn't happen

Reply:
`mp.amount` isn't properly updated when there is a new progressbar, because each worker has the state of `mp` when the task was spawned. They can't know if other workers have created new progressbars

I see two possible way of doing this
- use randomized ids for the progressbars, with low enough probability of collision, but I don't know how the interface `mp[i]` would work
- use some sort of second channel or another way to update about `mp.amount`

Reply:
I doubt that progressbars should be referred to using `mp[i]` at all, because of the possibility of race conditions and collisions as progressbars are added removed.

For that reason, maybe it's best to not use `push!` semantics (returning the collection), and instead stick with `p = addprogress!(mp, Type, args...)` returning the new progress bar (appended at the end)?

Why can't you do `p = addprogress!(mp, Type(args...))`, by the way?

Reply:
> Why can't you do p = addprogress!(mp, Type(args...)), by the way?

all functions actually do a `put!(channel, (progressid, action, args, kwargs))` for the main worker to process and apply the right function
`Progress[something]` created on one worker and called with `next!` on another doesn't work
this could be done with a function or a macro I guess
or create a new progress with all the same parameters

Reply:
> I doubt that progressbars should be referred to using mp[i] at all, because of the possibility of race conditions and collisions as progressbars are added removed.

when defining the progresses before, it makes sense to use `mp[i]`: (this works right now)

```julia
ntasks = 10
mp = MultipleProgress([Progress(100) for _ in 1:ntasks])
@sync for task in 1:ntasks
    @async for i in 1:100
        next!(mp[task])
        sleep(rand()/10)
    end
end
```

but when the progresses are defined in the task, it makes more sense like you say: (this doesn't work right now)

```julia
ntasks = 10
mp = MultipleProgress()
@sync for task in 1:ntasks
    p = addprogress!(mp, Progress, 100)
    @async for i in 1:100
        next!(p)
        sleep(rand()/10)
    end
end
finish!(mp)
```

I don't know if we should support both or if they should be 2 different structs

Reply:
@stevengj I don't really know how to implement this...

 I agree it would be nice to not have to specify an `id` and do 
```julia
p = addprogress!(...)
```

but we need to know the id of the newly created progress for `next!`, `update!` etc.

I had 3 ideas that all have issues:
- generate a random id (for example `rand(Int)`) but it can be easily broken with `@everywhere seed!(1)`. I don't know rng enough to know if this can be fixed with some sort of `rand(rng, Int)`
- have a second `RemoteChannel` so that the main worker gives the id back to the working worker. but I don't think we can make sure the right worker `take!`s the right id when 2 progresses are created at the same time
- include `myid()` in the id, but this will fail if a single worker spawns multiple progresses at the same time with `@async` for example